how-tos/providing-images-for-img.rst

```diff
@@ -0,0 +1,143 @@
+
+.. code-block:: nginx
+   server {
+     listen 80 default_server;
+     listen [::]:80 default_server;
+     server_name _;
+     root /var/www/html;
+     index index.html;
+     try_files $uri $uri/ =404;
+
+     location /api/ {
+       default_type  application/json;
+       proxy_hide_header Content-Location;
+       add_header Content-Location /api/$upstream_http_content_location;
+       proxy_set_header  Connection "";
+       proxy_http_version 1.1;
+       proxy_pass http://localhost:3000/;
+     }
+
+     location /files/ {
+       # /files/<id>/* ---> /files?select=blob&id=eq.<id>
+       rewrite /files/([^/]+).*  /files?select=blob&id=eq.$1  break;
+       # if id is missing
+       return 404;
+       # request binary output
+       proxy_set_header Accept application/octet-stream;
+       # usual proxy setup
+       proxy_hide_header Content-Location;
+       add_header Content-Location /api/$upstream_http_content_location;
+       proxy_set_header  Connection "";
+       proxy_http_version 1.1;
+       proxy_pass http://localhost:3000/;
+     }
+
+As you can see, we only explain the :code:`files` location.
+The reasoning for the rest of the recipe can be found elsewhere_.
+
+.. _elsewhere: ../admin.html#
+
+With this setup, we can request the cat image at :code:`localhost/files/42/cats.jpeg` without setting any headers.
+In fact, you can replace :code:`cats.jpeg` with any other filename or simply omit it.
+Putting the URL into the :code:`src` of an :code:`<img>` tag should now work as expected.
+
+Improved Version
+----------------
+
+The basic solution has some shortcomings:
+
+1.  The response :code:`Content-Type` header is set to :code:`application/octet-stream`.
+    This might confuse clients and users.
+2.  Download requests (e.g. Right Click -> Save Image As) to :code:`files/42` will propose :code:`42` as filename.
+    This might confuse users.
+3.  Requests to the binary endpoint are not cached.
+    This will cause unnecessary load on the database.
+
+The following improved version addresses these problems.
+First, we store the media types and names of our files in the database.
+
+.. code-block:: postgres
+   create table public.files(
+     id   int primary key
+   , type text
+   , name text
+   , blob bytea
+   );
+
+Next, we set up an RPC endpoint that sets the content type and filename.
+We use this opportunity to configure some basic, client-side caching.
+For production, you probably want to configure additional caches, e.g. on the reverse proxy.
+
+.. code-block:: postgres
+   set search_path=public
+
+   create function file(id int) returns bytea as
+   $$
+     declare headers text;
+     declare blob bytea;
+     begin
+       select format(
+         '[{"Content-Type": "%s"},'
+          '{"Content-Disposition": "inline; filename=\"%s\""},'
+          '{"Cache-Control": "max-age=259200"}]'
+         , f.type, f.name)
```

Comment:
I'm getting an error when running this function:
```
ERROR:  42P01: missing FROM-clause entry for table "f"
```
For solving this you could do:
```suggestion
         , type, name)
```